Parameter Sweeping for Benchmarking with R

davetcoleman opened the pull request:
moveit_benchmark_statistics.py
- Removed storing the planning request message in benchmark log files
- Added a new property of 'goal_name' for use with later results analysis
- Added an export to CSV function of benchmarking results for use with R, etc
- Added overwrite argument to easily reset the database file

benchmark_execution.cpp
- Added ability to do parameter sweeps
- Added more inline documentation
- Added new "workspace" and "workspace_frame" option to .cfg files for specifying the size of the planning workspace bounds
- Added new "sweep" option for specifying different methods of parameter sweeping during benchmarking
- Added new loop in benchmarking routine for changing parameters (sweeping)
- Modified the collected data to also work with different parameters, and to record them
- Created a modifyPlannerConfiguration() function

davetcoleman committed:
Added new command line arguments and ability to export all experiments to csv file

davetcoleman committed:
Added ability to store the goal name - the query, constraint, traj constraint, etc

davetcoleman committed:
Made debug output look better

davetcoleman committed:
Added parameter sweeping to benchmarking

davetcoleman committed:
More advanced parameter sweeping implmented, workspace bounds added

davetcoleman committed:
Changed for fractional factorial analysis

davetcoleman committed:
Merge branch 'groovy-devel' of github.com:davetcoleman/moveit_ros into groovy-devel

davetcoleman committed:
Code cleanup

davetcoleman commented:
Note: I removed the fractional factorial analysis functionality because it was hacked in last minute and not properly implemented. I have the code saved separately and could possibly re-add it in the future.

isucan commented on benchmarks/benchmarks/include/moveit/benchmarks/benchmark_execution.h:
```diff
@@ -79,6 +82,9 @@ struct BenchmarkRequest
   std::string filename;
+
+  // Optional: name of goal - only used for later analysis
+  std::string goal_name;
```
This is a member of a request. This should be a request name, not a goal name.

davetcoleman replied:
i named it goal name because its value is populated from the benchmark's .cfg file, from the "goal=" parameter. For example, it could hold the value "industrial.oven_01". i think its named appropriately...

isucan commented on benchmarks/benchmarks/include/moveit/benchmarks/benchmark_execution.h:
```diff
@@ -112,19 +118,74 @@ class BenchmarkExecution
+    // Parameters for sweeping
+    bool isSweep;
+    double start;
+    double iterator;
```
iterator has a different meaning. Perhaps step_size would be better?

davetcoleman replied:
fixed

isucan commented on benchmarks/benchmarks/include/moveit/benchmarks/benchmark_execution.h:
```diff
@@ -112,19 +118,74 @@ class BenchmarkExecution
+    std::string log_key;
+
+    // Parameters for sweeping
+    bool isSweep;
```
inconsistent naming. look at isSweep and log_key;

davetcoleman replied:
fixed

isucan commented on benchmarks/benchmarks/include/moveit/benchmarks/benchmark_execution.h:
```diff
@@ -112,19 +118,74 @@ class BenchmarkExecution
+  };
+
+  /// Allows for parameter sweeping of the planner configuration
+  struct FactorOptions
```
Why FactorOptions? Not ParameterOptions?

davetcoleman replied:
yea, i muddled the two terms. i was using this stuff for "fractional factorial analysis" and there they call it factors. they are synonyms, but i'll replace all the terms "factor" with "param"

isucan commented on benchmarks/benchmarks/include/moveit/benchmarks/benchmark_execution.h:
```diff
@@ -112,19 +118,74 @@ class BenchmarkExecution
+   * @brief Populates the factor_combinations_ vector with fractional factorial combination
+   * @return number of combinations to be tested
+   */
+  std::size_t generateFactorCombinationsHack();
```
hmm; this seems funky.

davetcoleman replied:
i removed it from the .cpp but forgot the header file, oops! this is my research code that i didn't get production ready :)

isucan commented on benchmarks/benchmarks/include/moveit/benchmarks/benchmark_execution.h:
```diff
@@ -112,19 +118,74 @@ class BenchmarkExecution
+   * @param factor_combinations_id_ - keeps track of what parameter combo we are currently iterating on
+   * @param parameter_data - used for outputting log information to file (results)
+   */
+  void modifyPlannerConfiguration(planning_interface::Planner* planner,
```
update instead of modify?

davetcoleman replied:
update would imply the previous configuration was stale or wrong, but in this case we are just changing it because we want the ability to test different param configurations. i vote leave it.

isucan commented on benchmarks/benchmarks/include/moveit/benchmarks/benchmark_execution.h:
```diff
@@ -112,19 +118,74 @@ class BenchmarkExecution
+  std::size_t generateFactorCombinations();
+
+  /// Recursively generates all the combinations of parameters to be tested
+  void recursiveFactorCombinations(int options_id, FactorInstance factor_instance);
```
What are option IDs?

davetcoleman replied:
documented.
side note: getting picky! i don't see other moveit code half as documented ;-)

isucan commented on benchmarks/benchmarks/src/benchmark_execution.cpp:
```diff
@@ -549,100 +613,155 @@ bool moveit_benchmarks::BenchmarkExecution::readOptions(const std::string &filen
+      else if (key.substr(0, 6) == "sweep.")
+      {
+        std::string sweep_var = key.substr(6);
+        ROS_ERROR_STREAM_NAMED("temp","SWEEP key="<<sweep_var<<" val="<<val);
```
do we need this?

davetcoleman replied:
totally not. i just removed a bunch of debug output. sorry this code is not better cleaned up, but its been 2 months and i'm distracted by lots of other projects. i just wanted to make sure it didn't get merged!

isucan commented on benchmarks/benchmarks/src/benchmark_execution.cpp:
```diff
@@ -549,100 +613,155 @@ bool moveit_benchmarks::BenchmarkExecution::readOptions(const std::string &filen
+        // Insert into array of all sweeps
+        factor_options_.push_back(new_sweep);
+
+        //ROS_ERROR_STREAM_NAMED("temp","SWEEP start " << new_sweep.start << " it " << new_sweep.iterator << " end " << new_sweep.end);
```
can we remove this?

davetcoleman replied:
done

isucan commented:
Does this set of changes affect how previously implemented benchmarks run normally?

davetcoleman committed:
Merge branch 'groovy-devel' of https://github.com/ros-planning/moveit_ros into groovy-devel

davetcoleman committed:
Fixed per Ioan's code review

davetcoleman committed:
Fixed for moveit_msgs/JointLimits.h no such file or directory

davetcoleman commented:
I just added another [commit](https://github.com/davetcoleman/moveit_ros/commit/4abf30cfcfa2939e7003454183b0421df297b78c) that is unrelated to this pull request but automatically got sucked in, oops:

Fixed for moveit_msgs/JointLimits.h no such file or directory. This is the error msg I was getting:

```
[  3%] Generating Python from MSG moveit_msgs/DisplayTrajectory
In file included from /home/dave/ros/ws_moveit/src/moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h:40:0,
                 from /home/dave/ros/ws_moveit/src/moveit_core/robot_model/include/moveit/robot_model/robot_model.h:45,
                 from /home/dave/ros/ws_moveit/src/moveit_core/planning_scene/include/moveit/planning_scene/planning_scene.h:40,
                 from /home/dave/ros/ws_moveit/src/moveit_core/planning_interface/include/moveit/planning_interface/planning_interface.h:40,
                 from /home/dave/ros/ws_moveit/src/moveit_ros/planning/planning_pipeline/include/moveit/planning_pipeline/planning_pipeline.h:40,
                 from /home/dave/ros/ws_moveit/src/moveit_ros/move_group/src/move_group_context.cpp:39:
/home/dave/ros/ws_moveit/src/moveit_core/robot_model/include/moveit/robot_model/joint_model.h:44:37: fatal error: moveit_msgs/JointLimits.h: No such file or directory
compilation terminated.
```

amdshameer commented:
@davetcoleman What exactly does this sweep do? For ex:sweep.range=1:10:100, Is it setting 10 values between 1 to 100 to the parameter we specify and run the same experiment for those ten values? If so, the log file should have 10 set of runs?
